cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -670,26 +833,36 @@ std::shared_ptr<ScalarFunction> MakeTemporal(
   auto func =
       std::make_shared<ScalarFunction>(name, Arity::Unary(), doc, default_options);
 
+  {
```

Comment:
Frankly I think kernels like `hour` should just not work on dates - it's not meaningful.

Reply:
Make sense, pandas doesn't has that behavior too
```python
pd.to_datetime("2019-01-03", format='%Y-%m-%d', errors='coerce').date().hour()
AttributeError: 'datetime.date' object has no attribute 'hour'
```
I'll remove the support of date32/64 for hour, min, sec ...